src/cpp/include/aiebu/aiebu_assembler.h

```diff
@@ -152,53 +164,81 @@ class aiebu_assembler
      * @class argtbl
      *
      * @brief
-     * aiebu_assembler::argtbl represents a table of xrt_idx and BD offset.
-     * The table is constructed from an aiebu_assembler object based on the
-     * control code that indicate which xrt_idx should patched into which BD.
+     * aiebu_assembler::argtbls represents a vector of instance infor.
+     * Inside each element, there is a table of xrt_idx and BD offset for
+     * that instance.
+     *
+     * The class is constructed from an aiebu_assembler object based on the
+     * control code that indicate which xrt_idx should patched into which BD
+     * for each instance within the given kernel name.
      *
      * This object can be used to dump the table and modify the xrt_idx and its
-     * BD offset in any entry from an aiebu_assembler object. And then flush it
+     * BD offset in any entry in any instance in a given kernel. And then flush it
      * back to aiebu_assembler so that the xrt_id and its BD offset can be updated
      * in the control code and .dynamic sections of ELF.
      *
+     * The kernel name can be updated at the same time by calling set_name() API.
+     * So that when this object is flushed back, the kernel name can be updated in
+     * ELF as well.
+     *
      * When using this class to do xrt argument transform
      *      1. Only host patching is supported.
      *      2. Only support AIE2PS and AIE4
      */
-    class argtbl_impl;
-    class argtbl
-    {
-      private:
-        std::shared_ptr<argtbl_impl> handle;
-      public:
-        explicit argtbl(std::shared_ptr<argtbl_impl> in_impl);
-
-        /*
-         * Dump the reference of vetor of xrt argument (xrt_idx)
-         * and BD offset. Caller can modify entries in the vector
-         * in place. Then the whole argtbl object can be flushed
-         * back to aiebu_assembler to update the ELF to do the
-         * xrt_idx transform.
-         */
-        std::vector<arginfo>& dump() const;
-
-    };
-    /*
-     * Get an argtbl object from aiebu_assembler. In this function
-     * aiebu_assember will scan the ELF and construct a table (vector)
-     * of xrt_idx and BD offset. Then it returns an object of this table
-     * which can be used to dump the reference of the table and modify
-     * the table in place.
-     */
-    argtbl get_argtbl();
-
-    /*
-     * Flush the argtbl object to aiebu_assembler. In this function,
-     * aiebu will take the argtbl object and update control code and
-     * patching metadata based on the xrt_idx and BD offset in the
-     * arg table.
-     */
-    void flush_argtbl(const argtbl& arg_table);
+     class argtbl
+     {
+         friend class aiebu_assembler;
```

Comment:
why do we need this?

Reply:
Handle classes can safely return the handle through a public inline API (get_handle()).  The handle is useless to public clients, but Implementation code can then access the impl this way.   It accomplishes the same as the friend class though, so really no preference one way or the other.

FWIW, XRT uses get_handle for all pimpls, see here https://github.com/Xilinx/XRT/blob/5c6def2df468485b7240e3749d94994f0deeeb2f/src/runtime_src/core/include/xrt/detail/pimpl.h#L52